Integrate Calico network plugin with vagrant provisioner

Symmetric opened the pull request:
I've been working on a prototype integration of [Calico](https://github.com/Metaswitch/calico-docker) networking in Kubernetes. I've used the new [network plugin API](https://github.com/GoogleCloudPlatform/kubernetes/pull/5069) to set up the Pod infra container with Calico networking, and the basic use-case is working.

This adds tenant isolation to Kubernetes, so that pod traffic can be restricted using the Calico distributed firewall. It also means that network traffic between nodes is unencapsulated, since Calico uses an L3 BGP network rather than an L2 overlay network like OVS/flannel.

At this point I'd like to get some feedback on the general approach, to see if there's anything that I should be doing differently.
# Overview

Here's an overview of the structure:
- Add a NETWORK_MODE to chose between openvswitch and calico.
- Update the salt files to install the Calico infrastructure container on each minion.
- Update the salt files to run a new etcd instance on the master, listening on port 6666
  - Needed for the Calico etcd datastore, which can't be run as a pod since we need Calico to be up before we can start pods.
- On pod creation:
  - Record the existing (docker-allocated) pod IP.
  - Delete the bridged veth.
  - Add a new veth pair with the docker-allocated Pod IP, using Calico networking. The node-end of the veth pair is not attached to a vSwitch or docker bridge, instead it is processed by the Linux routing table (i.e. a vRouter).
# Questions

This works fine with direct Pod-Pod communication, but I'm not sure about Services -- in order to firewall a flow properly, we need to see the real source IP when it hits the iptables rules on the destination node. I'm concerned that the Service is going to get in the way of this; it looks like the service proxy is rewriting the source IP of packets to stay in the reverse flow, is that correct?

If so, any suggestions for how we could get the Service proxy to play nice with Calico?

And, some general questions on the architecture described above:
- Should I move the calico-node container initialization from the salt files to the plugin's `setup` step?
  - I use salt config for setting up the calico node, so not sure what's preferable there.

Any other feedback on the design is welcome.

CC: @thockin @lxpollitt
# TODO
- [x] Add `setup` handler in network plugin
- [x] Add `teardown` handler in network plugin
- [x] Set default policy profile in Calico plugin, so that pods have global connectivity by default
- [x] Tidy up OVS/calico flags
- [x] Figure out how Services are going to work (wait for https://github.com/GoogleCloudPlatform/kubernetes/issues/6666 and/or https://github.com/GoogleCloudPlatform/kubernetes/pull/9210)

Symmetric commented:
After digging into the kube-proxy a bit more, I'm pretty sure we're not going to be able to do firewalling with the current Kubernetes implemenatation.

If a client sends a packet to a service IP, then the service pod is going to see a request from an IP owned by kube-proxy. This means that we can't do inbound filtering at the pod, since we don't know the client's true source address/port. Outbound filtering at the client's end is unwieldy here, though possible with some extensions; we could watch the API for new services, and install outbound firewall rules blocking access to certain services. But this isn't as effective as inbound filtering at the pod, since it doesn't protect against compromised endpoints on the network.

Looks like we need something like DNS SRV (or multiple A records), as suggested by https://github.com/GoogleCloudPlatform/kubernetes/issues/6666. In that case the client would communicate directly with the pod, and our firewall would work at both the traffic source and destination.

Symmetric commented:
I signed it! (as Metaswitch Networks).

On Thu, Apr 23, 2015 at 1:00 PM, googlebot <notifications@github.com> wrote:

> Thanks for your pull request. It looks like this may be your first
> contribution to a Google open source project, in which case you'll need to
> sign a Contributor License Agreement (CLA).
>
> [image: :memo:] *Please visit https://cla.developers.google.com/
> <https://cla.developers.google.com/> to sign.*
>
> Once you've signed, please reply here (e.g. I signed it!) and we'll
> verify. Thanks.
> ------------------------------
>
>    - If you've already signed a CLA, it's possible we don't have your
>    GitHub username or you're using a different email address. Check your
>    existing CLA data <https://cla.developers.google.com/clas> and verify
>    that your email is set on your git commits
>    <https://help.github.com/articles/setting-your-email-in-git/>.
>    - If you signed the CLA as a corporation, please let us know the
>    company's name.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7245#issuecomment-95702204>
> .
>

Commit:
Add Calico network plugin for Vagrant provisioner.

Adds support for Calico networking in Kubernetes. Nodes are now
joined in a BGP mesh. When a pod is created, a new route for that
pod is distributed to all other Kubernetes nodes. This means
that network traffic between all containers on the network is
un-encapsulated (no need for an overlay network).

Calico also provides ACL function, which may be configured using
the calicoctl tool (or programmed directly into the etcd
datastore). This allows intent-based firewalling at both the
source and destination pod.

Commit:
Update calico plugin to disable firewalling.

For first integration attempt, we want to disable the firewall
function and preserve the existing kubernetes access model
(that being no restrictions of any sort).

Commit:
Pin calico node version.

Commit:
Bugfixes for stability

Commit:
Delete profile when pod is removed.

Commit:
Properly remove profile on pod delete.

Previously was just removing container from profile on pod delete,
so was leaking profiles. Now delete the profile too.

Commit:
Upgrade calico to 0.4.2.

Move calico-network-plugin code to calico-docker repo, and
link directly to the binary from the salt files.

No need to store the calico plugin in the Kubernetes tree now.

Commit:
Merge branch 'master' into calico-network-plugin-dev

Conflicts:
	cluster/saltbase/salt/kubelet/default

Commit:
Fix salt bug by pinning version.

https://github.com/GoogleCloudPlatform/kubernetes/issues/8114

Commit:
Complete update to calico/node v0.4.2

Commit:
Fix salt upgrade bug by defensively importing.

Fix for https://github.com/GoogleCloudPlatform/kubernetes/issues/8114

Commit:
Merge branch 'calico-network-plugin-dev-workaround' into calico-network-plugin-dev

Symmetric commented:
cc @spikecurtis

Commit:
Remove code added for old salt fix, and DNAT change.

Revert the change to the kube-proxy to avoid REDIRECT NAT,
and remove some orphaned shell code for vagrant setup.

lxpollitt commented:
Note that I'm not sure why this is labeled "cla: no" since the contributors license agreement has been signed and submitted.

jdef commented:
Would the proxy still be in the way (appearing as "the client") if everything was done via iptables instead of userspace proxy impl?
- https://github.com/GoogleCloudPlatform/kubernetes/pull/9210
  - https://github.com/GoogleCloudPlatform/kubernetes/issues/3760

spikecurtis commented:
@jdef #9210 should preserve the actual pod IP as the source when the packets hit the service instances, which would allow our firewall to operate correctly (yay!).

It may also alleviate the difficulty on older kernels of `REDIRECT` versus `DNAT` rules that currently forward traffic to the proxy (`REDIRECT` rules don't work for Calico if the proxy is not bound to a specific IP address on the kubelet).  I'd need to check with @Symmetric to be sure about what testing he's done in this area.

BenTheElder commented:
Note that when looking at #9210 the iptables rules as generated now may not be entirely correct, I'll be working to finish checking that once vagrant is working again. We also haven't selected a target iptables minimum version yet.

Any comments or suggestions on the rules used etc would be appreciated, I'd like to get this right the first time, and I'd like to make it play nice with the work here.

Commit:
Merge branch 'master-for-ben' into calico-network-plugin-dev

Conflicts:
	cluster/saltbase/salt/docker/init.sls
	cluster/saltbase/salt/kubelet/default
	cluster/saltbase/salt/top.sls

Commit:
Update calico from 0.4.2 to 0.4.8

Update versions of calico-docker hosted files.

Commit:
Tidy up CIDR env variables.

Reduce the number of changes to the ovs Vagrant mode by
more careful choice of the ENV variables that we export.

Commit:
Revert changes to docker-defaults

We don't use this file any more, so just leave it as it is.

salt/docker/default is now where both OSs get their defaults from.

Commit:
Add NETWORK_MODE env variable to Vagrant scripts

Allow calico to be enabled using a NETWORK_MODE=calico env
variable, instead of replacing OVS.

Commit:
Fix typo in docker init.sls

Commit:
Propagate NETWORK_MODE into nodes

Wasn't applying the NETWORK_MODE env variable onto the nodes
that Vagrant creates.

Commit:
Create cbr0 docker bridge on master node.

Update salt files to create cbr0 docker bridge on the master node.

Used to do this, but merging latest master must have broken that.

Commit:
Update calico_kubernetes plugin in 0.4.8 release.

Replaced the plugin file on Github, so need to update the sha
hash in the salt files.

Commit:
Merge branch 'master' into calico-network-plugin-dev

Conflicts:
	cluster/saltbase/salt/_states/container_bridge.py
	cluster/saltbase/salt/kubelet/default
	cluster/saltbase/salt/top.sls

Commit:
Add code to setup Calico cbr0 for docker.

Symmetric commented:
Some fairly major refactoring has just gone in. I've rebased to the latest master, which has required some change to the structure (e.g. how/where we set up the cbr0 bridge). Functionally I think we're good, but there are a couple extraneous bash variables still lying around.

Also, I've moved the plugin code onto the calico-docker repository. It's versioned, so new builds there won't break what's checked in here.

paultiplady committed:
Revert rearrangement of docker/init.sls

thockin commented:
I finally took a look at this.  Overall it's in-line with the current codebase, so I can't complain, but I will anyway.  The salt is becoming a real tangle of options that are impossible to ensure work in all cases.  As cross through the 1.0 gate we need to think hard about how we're going to a) make this sort of thing more self-contained, b) make this testable, c) offload the testing to sub-owners to ensure (e.g.) Calico keeps working without our own test machines running e2e N-times on every PR.

paultiplady committed:
Final code review tweaks to salt files.

paultiplady committed:
Improve logging in provision-network.sh

paultiplady committed:
Merge branch 'master' into calico-network-plugin-dev

Conflicts:
	cluster/vagrant/config-default.sh

paultiplady committed:
Update calico_kubernetes plugin hash.

Symmetric commented:
@thockin thanks for taking a look. I agree that the salt files are pretty tangled; it took a lot more effort than expected to get this working. I have been pondering what it would look like in an ideal world, and I have a few ideas for possible improvements.

I think that it would be worth exploring the possibility of pulling out the provisioning providers in kubernetes into separate repos, rather than checking them into the main k8s repo. The Vagrant provider model is a good one; if structured correctly, externally-hosted providers could be importable by the main k8s cluster tool. That way a standard UI for cluster operations could be maintained, without requiring all provider code to live in the k8s repo.

I'm up for spending some time kicking ideas around and working on kube-provisioning 2.0, and refactoring the Calico provider to the New Way. I imagine there will be a very wide audience for that discussion though.

Would you be happy to take this PR, and address the more fundamental questions of provisioning plugins once the dust settles from 1.0?

paultiplady committed:
Merge tag 'v0.21.1' into calico-network-plugin-dev

Kubernetes version v0.21.1

Conflicts:
	cluster/vagrant/config-default.sh
	cluster/vagrant/provision-master.sh
	cluster/vagrant/util.sh

paultiplady committed:
Run the calico-node daemon on the Kube master.

Need this to distribute routes for the pods.

paultiplady committed:
Merge tag 'v1.0.0' into calico-network-plugin-dev

Kubernetes version v1.0.0

paultiplady committed:
Run calico daemon on the k8s master.

Required to properly distribute routes for the pod IPs onto the
master. (Was previously working by some unknown magic).

piosz commented:
Is this PR still active?

caseydavenport commented:
I believe @Symmetric was still hoping to get this in, but was waiting for some more feedback / direction from @thockin.  It looks like this needs a rebase, and @Symmetric will likely be inactive until Monday.  We can wait to see what he has to say on it, or close it for now and we can re-open when he gets back.

thockin commented on cluster/vagrant/provision-master.sh:
```diff
@@ -71,8 +71,15 @@ done
 
-# Configure the openvswitch network
-provision-network
+if [ $NETWORK_MODE != "calico" ]; then
```
I would make this explicitly == "openvswitch"

thockin commented on cluster/vagrant/provision-master.sh:
```diff
@@ -71,8 +71,15 @@ done
+  # Configure the openvswitch network
+  echo "Provisioning openvswitch network"
+  provision-network
+else
```
else if "calico"

thockin commented on cluster/vagrant/provision-master.sh:
```diff
@@ -71,8 +71,15 @@ done
+  # Configure the cbr for Calico networking.
+  echo "Provisioning cbr0 for Calico"
+  provision-network-calico
+fi
```
else abort

thockin commented:
I'm OK with this.  My complaints from before stand, but they are not your fault.

@derekwaynecarr for final LGTM for vagrant

jdef commented:
/cc @karlkfi

paultiplady committed:
Explicitly check for known NETWORK_MODE for nodes

Rather than have !calico and calico, instead enumerate the
known network modes (calico, openvswitch) and error if we get
something that's unexpected.

paultiplady committed:
Merge branch 'master' into calico-network-plugin-dev

Conflicts:
	cluster/saltbase/salt/kubelet/default
	docs/salt.md

Symmetric commented:
Thanks for taking a look.

Review markups applied in 4c580f4, want me to squash now or wait for final ack?

Symmetric commented:
@derekwaynecarr -- any chance you can take a look at this? Would be great to minimize the pain of rebasing and get this in...

Just been resized as XS, so shouldn't take much of your time. Cheers!

kelseyhightower commented:
@Symmetric Hello! I promise if you rebase this again I'll spend the time to finish up the review. Also thanks for taking the time to put together this PR.

Symmetric commented:
Cheers Kelsey, I'll rebase this week, hopefully later today.

paultiplady committed:
Merge branch 'master' into calico-network-plugin-dev-master-1.2

Conflicts:
	cluster/saltbase/salt/kubelet/default
	cluster/saltbase/salt/top.sls
	cluster/vagrant/provision-master.sh
	cluster/vagrant/provision-minion.sh
	cluster/vagrant/provision-network.sh
	cluster/vagrant/util.sh

paultiplady committed:
Update master & node to calicoctl 0.10.0

paultiplady committed:
Fix merge issues

paultiplady committed:
Merge branch 'master' into calico-network-plugin-dev-master-1.2

paultiplady committed:
Remove calico network provisioning step

Not needed now that we have Calico IPAM.

paultiplady committed:
Restructure top.sls to allow Calico on all clouds

paultiplady committed:
Use plugin config file

Use a config file to store the calico-kubernetes plugin
config (specifically, the auth token).

paultiplady committed:
Bugfixes in calico_kubernetes.ini

paultiplady committed:
Manually download the calico-kubernetes plugin

paultiplady committed:
Ensure correct pools when kubelet starts

paultiplady committed:
Upgrade plugin to v0.6.1

paultiplady committed:
Fix pool race with kubelet

Kubelet was starting with the 192.168 pool, even though
there was a require: service.kubelet dependency.

Instead, depend on the kubelet binary file itself.

paultiplady committed:
Add --nat-outgoing to provide network connectivity

Without the --nat-outgoing flag, we have no internet connectivity
in pods, because nodes outside the k8s cluster don't know about
the pod IP ranges.

Symmetric commented:
@kelseyhightower this turned out to be more involved than I thought, but it's working now; PTAL.

@caseydavenport will be taking this over the finish line.

Cheers!

derekwaynecarr commented:
@Symmetric @caseydavenport - please poke the PR when its ready again.

also please update with a note in here:
https://github.com/kubernetes/kubernetes/tree/master/docs/getting-started-guides

that you will provide future support for calico based network deployment issues for the affected platforms.

Thanks!

dcbw commented:
@Symmetric this and https://github.com/kubernetes/kubernetes/pull/18795 are going to clash a bit, so whichever of us merges first we'll have to sort that out.

lxpollitt commented:
@caseydavenport: see @dcbw's comment above and please coordinate with him

caseydavenport commented:
@dcbw - Yep.  I think this PR is already quite out-of-date as it is.  I'd suggest you go first, and I'll worry about this guy afterwards.

I think we can close this PR, and I can open up another when I get around to rebasing, etc.
